File: remote/src/main/java/com/kelsos/mbrc/services/ProtocolHandler.java
package com.kelsos.mbrc.services;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.kelsos.mbrc.BuildConfig;
import com.kelsos.mbrc.constants.Const;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.constants.ProtocolEventType;
import com.kelsos.mbrc.events.MessageEvent;
import com.squareup.otto.Bus;
import java.io.IOException;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.map.ObjectMapper;
import roboguice.util.Ln;

@Singleton public class ProtocolHandler {
  private Bus bus;
  private boolean isHandshakeComplete;
  private ObjectMapper mapper;

  @Inject public ProtocolHandler(Bus bus, ObjectMapper mapper) {
    this.bus = bus;
    this.mapper = mapper;
  }

  public void resetHandshake() {
    isHandshakeComplete = false;
  }

  public void preProcessIncoming(final String incoming) {
    try {
      final String[] replies = incoming.split("\r\n");
      for (String reply : replies) {

        JsonNode node = mapper.readValue(reply, JsonNode.class);
        String context = node.path("context").getTextValue();

        if (context.contains(Protocol.ClientNotAllowed)) {
          bus.post(new MessageEvent(ProtocolEventType.InformClientNotAllowed));
          return;
        }

        if (!isHandshakeComplete) {
          if (context.contains(Protocol.Player)) {
            bus.post(new MessageEvent(ProtocolEventType.InitiateProtocolRequest));
          } else if (context.contains(Protocol.ProtocolTag)) {
            isHandshakeComplete = true;
            bus.post(new MessageEvent(ProtocolEventType.HandshakeComplete, true));
          } else {
            return;
          }
        }

        bus.post(new MessageEvent(context, node.path(Const.DATA)));
      }
    } catch (IOException e) {
      if (BuildConfig.DEBUG) {
        Ln.d("Incoming preprocessor", e);
        Ln.d("While processing: %s", incoming);
      }
    }
  }
}

File: app/src/main/java/com/kelsos/mbrc/networking/protocol/ProtocolHandler.kt
package com.kelsos.mbrc.networking.protocol

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.kelsos.mbrc.common.state.MainDataModel
import com.kelsos.mbrc.constants.Const
import com.kelsos.mbrc.constants.ProtocolEventType
import com.kelsos.mbrc.events.MessageEvent
import com.kelsos.mbrc.events.bus.RxBus
import timber.log.Timber
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ProtocolHandler
  @Inject
  constructor(
    private val bus: RxBus,
    private val mapper: ObjectMapper,
    private val model: MainDataModel,
  ) {
    private var isHandshakeComplete: Boolean = false

    fun resetHandshake() {
      isHandshakeComplete = false
    }

    fun preProcessIncoming(incoming: String) {
      try {
        val replies =
          incoming
            .split("\r\n".toRegex())
            .dropLastWhile(String::isEmpty)
            .toTypedArray()

        replies.forEach {
          Timber.Forest.v("received:: $it")

          val node = mapper.readValue(it, JsonNode::class.java)
          val context =
            node
              .path("context")
              .textValue()
              .trim()
              .lowercase(Locale.getDefault())

          if (context == Protocol.CLIENT_NOT_ALLOWED) {
            bus.post(MessageEvent(ProtocolEventType.INFORM_CLIENT_NOT_ALLOWED))
            return
          }

          if (!isHandshakeComplete) {
            when (context) {
              Protocol.PLAYER -> bus.post(MessageEvent(ProtocolEventType.INITIATE_PROTOCOL_REQUEST))
              Protocol.PROTOCOL_TAG -> handleProtocolMessage(node)
              else -> return
            }
          }

          bus.post(MessageEvent(context, node.path(Const.DATA)))
        }
      } catch (e: Exception) {
        Timber.Forest.v(e, "Failure while processing incoming data")
      }
    }

    private fun handleProtocolMessage(node: JsonNode) {
      model.pluginProtocol = getProtocolVersion(node)
      if (model.apiOutOfDate) {
        bus.post(MessageEvent(ProtocolEventType.PLUGIN_UPDATE_AVAILABLE))
      }
      isHandshakeComplete = true
      bus.post(MessageEvent(ProtocolEventType.HANDSHAKE_COMPLETE, true))
    }

    private fun getProtocolVersion(node: JsonNode): Int =
      try {
        Integer.parseInt(node.path(Const.DATA).asText())
      } catch (ignore: Exception) {
        2
      }
  }
